sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/runtime/pmp.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/stdasm.h"
+
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6.1 Physical Memory Protection CSRs",
+// "Figure 3.28: PMP configuration register format".
+#define PMP_CFG_CSR_R 0
+#define PMP_CFG_CSR_W 1
+#define PMP_CFG_CSR_X 2
+#define PMP_CFG_CSR_A 3
+#define PMP_CFG_CSR_L 7
+
+#define PMP_CFG_FIELDS_PER_REG 4
+#define PMP_CFG_FIELD_WIDTH 8
+#define PMP_CFG_FIELD_MASK 0xffff
+
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6 Physical Memory Protection", "Address Matching".
+#define PMP_CFG_CSR_MODE_OFF 0
+#define PMP_CFG_CSR_MODE_TOR 1
+#define PMP_CFG_CSR_MODE_NA4 2
+#define PMP_CFG_CSR_MODE_NAPOT 3
+#define PMP_CFG_CSR_MODE_MASK 0x3
+
+// TODO - explanation to why macro magic is necessary (csr must be const expr).
+#define PMP_CSR_WRITE(csr, var) asm volatile("csrw " #csr ", %0;" : : "r"(var))
```

Comment:
Also, a separate thought: as written, this code cannot be unit-tested. We should think about maybe using @alphan and I's "Mask ROM component mocks" as a potential model for mocking CSR IO?

This is for a followup, obviously, but I wanted to start thinking about it.

Reply:
I think it can be unittested through `pmp_cfg_csr_rw` and `pmp_address_csr_rw`, bare in mind these functions are subject to change as per @lenary X-macros pattern suggestion.

Reply:
You can't mock partially mock a C file with our mocking system, unfortunately, so we probably can't do it that way.